okio/jvm/src/main/java/okio/Options.kt

```diff
@@ -37,28 +37,34 @@ class Options private constructor(
       // Sort the byte strings which is required when recursively building the trie. Map the sorted
       // indexes to the caller's indexes.
       val sortedByteStrings = byteStrings.sortedArray()
-      // TODO(jwilson): this is a O(N*N) loop to recover the original indexes. Ick.
-      val sortedIndexes = IntArray(sortedByteStrings.size) {
-        byteStrings.indexOf(sortedByteStrings[it])
+      val sortedIndexes = IntArray(sortedByteStrings.size)
+      byteStrings.forEachIndexed { callerIndex, byteString ->
+        val sortedIndex = sortedByteStrings.binarySearch(byteString)
+        sortedIndexes[sortedIndex] = callerIndex
       }
       require(sortedByteStrings[0].size() > 0) { "the empty byte string is not a supported option" }
 
       // Strip elements that will never be returned because they follow their own prefixes. For
       // example, if the caller provides ["abc", "abcde"] we will never return "abcde" because we
       // return as soon as we encounter "abc".
-      val trieByteStrings = mutableListOf<ByteString>()
-      val trieIndexes = mutableListOf<Int>()
-      collectEligible@
-      for (i in 0 until sortedByteStrings.size) {
-        val byteString = sortedByteStrings[i]
-        val index = sortedIndexes[i]
-        for (j in trieIndexes.size - 1 downTo 0) {
-          if (byteString.startsWith(trieByteStrings[j]) && index >= trieIndexes[j]) {
-            continue@collectEligible // This value can never be returned.
+      val trieByteStrings = sortedByteStrings.toMutableList()
+      val trieIndexes = sortedIndexes.toMutableList()
```

Comment:
`sortedIndexes` can be declared as `MutableList` from the beginning to avoid creating a copy. Same for `sortedByteStrings`, we can sort `byteStrings` in place and copy to a `MutableList` - currently it's being copied twice.